src/jsonrpc.rs

```diff
@@ -0,0 +1,62 @@
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+
+#[derive(Debug, PartialEq, Clone, Hash, Eq, Deserialize, Serialize, PartialOrd, Ord)]
+#[serde(deny_unknown_fields)]
+#[serde(untagged)]
+pub enum Id {
+    /// Null
+    Null,
+    /// Numeric id
+    Number(u64),
+    /// String id
+    Str(String),
+}
+
+/// JSON-RPC request object as defined in the [spec](https://www.jsonrpc.org/specification#request-object).
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct Request<T> {
+    /// JSON-RPC version.
+    pub jsonrpc: String,
+    /// Request ID
+    pub id: Option<Id>,
+    /// Name of the method to be invoked.
+    pub method: String,
+    /// Parameter values of the request.
+    pub params: Option<T>,
+}
+
+/// JSON-RPC Response object as defined in the [spec](https://www.jsonrpc.org/specification#request-object).
+/// TODO: Figure out a cleaner way to do this.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+#[serde(untagged)]
+pub enum Response<T, E> {
```

Comment:
This looks very similar to the builtin `Result<T, E>` type.

Reply:
Yep, I tried adding a From/Into pair, but it was awkward with the extra "id" field.